mathesar_ui/src/component-library/common/utils/pauseableTweened.ts

```diff
@@ -0,0 +1,95 @@
+import type { Tweened } from 'svelte/motion';
+import { tweened } from 'svelte/motion';
+import { get } from 'svelte/store';
+
+type Updater<T> = (target_value: T, value: T) => T;
+
+/** Copied from tweened.d.ts */
+interface TweenedOptions<T> {
+  delay?: number;
+  duration?: number;
+  easing?: (t: number) => number;
+  interpolate?: (a: T, b: T) => (t: number) => T;
+}
+
+interface Options extends TweenedOptions<number> {
+  allowPause: boolean,
+}
+
+const defaults = {
+  allowPause: true,
+  duration: 0,
+};
+
+export interface PauseableTweened extends Tweened<number> {
+  pause: () => void,
+  resume: () => void,
+}
+
+export function pauseableTweened(
+  initialValue: number,
+  options: Partial<Options> = {},
+): PauseableTweened {
+  const fullOptions = { ...defaults, ...options };
+  const store = tweened(initialValue, fullOptions);
+
+  let isPaused = false;
+  let targetValue = initialValue;
+  let startingValue = initialValue;
+  let { duration } = fullOptions;
+  let promiseResolve = () => {};
+
+  function makePromise() {
+    return new Promise<void>((resolve) => { promiseResolve = resolve; });
+  }
+
+  function set(value: number, opts?: TweenedOptions<number>): Promise<void> {
+    targetValue = value;
+    duration = opts?.duration ?? fullOptions.duration;
+    const promise = makePromise();
+    void store.update((_targetValue, _currentValue) => {
+      startingValue = _currentValue;
+      return value;
+    }, opts).then(promiseResolve);
+    return promise;
```

Comment:
Is there a reason why we create a new promise that sets `promiseResolve` instead of directly returning the promise which is returned from `store.update`?

Reply:
Yeah good question.

This `PauseableTweened` store was a little tricky to create. The code would probably be more straightforward if I were to write it from scratch. Instead, I built it on top of Svelte's `Tweened` store because that seemed faster, if albeit somewhat more hacky.

In `ToastController.show` we create a `PauseableTweened` store for `controller.progress` and then call:

```ts
controller.progress.set(props.finalProgress, { duration: props.lifetime })
  .then(controller.dismiss);
```

For that code to work correctly, the Promise retuned by `PauseableTweened.set` needs to resolve when the timer runs out -- even if the user has paused and resumed the timer. The problem is that, with the way `Tweened` works, the only mechanism I have to pause the timer is to issue second `Tweened.set` call (in addition to the first `Tweened.set` that initiated the timer). Onece a second `Tweened.set` is issued to pause the timer, the Promise returned from the first `Tweened.set` will never resolve, meaning that the `controller.dismiss` function will never run.

Let's say we changed `PauseableTweened.set` to the following (as I gather you're suggesting):

```ts
async function set(value: number, opts?: TweenedOptions<number>): Promise<void> {
  targetValue = value;
  duration = opts?.duration ?? fullOptions.duration;
  return store.update((_targetValue, _currentValue) => {
    startingValue = _currentValue;
    return value;
  }, opts);
}
```

With that code, the toast message would work fine in most cases, but if the user hovers on the message to pause it, then the toast message will not automatically close itself when the timer runs out afterwards.

If you can think of a better way to structure this code, I'm all for it. Or I could add some code comments atop `PauseableTweened.set` which explain this a bit.

Reply:
Sounds good. I took a second look with the reason in mind and the usecase makes sense. I think it would be better to add some comments.

Reply:
I added comments in e4c0fbf63d4249d3fcca7ad9c724bc3ebd63bbf7 that I hope should resolve this